### ISSUE TITLE
Avoid generating a body for HEAD requests in `twisted/web/_auth/wrapper.py`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5490]
|<img alt="garethbult's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| garethbult reported|
|-|-|
|Trac ID|trac#5490|
|Type|enhancement|
|Created|2012-02-18 00:02:19Z|
|Branch|https://github.com/twisted/twisted/tree/head-unauthorized-5490|

The default render function in `twisted/web/_auth/wrapper.py` that's called to handle http authentication returns an 'Unauthorized' string. If used in combination with a HEAD request, this results in the return from the head request containing a body, even though at application this may have been left empty.

In web/server.py on line 227, this is picked up resulting in the server logging "HEAD request for resource, I think I'll eat it".

My solution is to return `''` rather than 'Unauthorized' on line 57 of _auth/wrapper.py.

<details><summary>Searchable metadata</summary>

```
trac-id__5490 5490
type__enhancement enhancement
reporter__garethbult garethbult
priority__normal normal
milestone__ 
branch__branches_head_unauthorized_5490 branches/head-unauthorized-5490
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1329523339000000 1329523339000000
changetime__1329669789000000 1329669789000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Thanks.  I think you're talking about `twisted/web/_auth/wrapper.py` - please do use full paths to filenames, though.

Also, the `Twisted-12.0` milestone is over, so no new tickets should be added to it (in fact, it should have been closed as part of the Twisted 12 release.  I'm not sure why it wasn't).  In general, the release milestones are for regressions or other critical tickets, not normal bug fixes or feature requests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [33569]) Branching to 'head-unauthorized-5490'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/head-unauthorized-5490)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Looks good to me, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [33575]) Merge head-unauthorized-5490 

Author: exarkun
Reviewer: therve
Fixes: [#5490](https://github.com/twisted/twisted/issues/5490)

When generating an unauthorized response from `twisted.web.guard.SessionWrapper`,
only write to the response body if the request method is not HEAD.  This avoids
generating a warning in the log but should not otherwise change the behavior
of `SessionWrapper`.
